src/stan-users-guide/cross-validation.Rmd

```diff
@@ -0,0 +1,390 @@
+carried out, the first of which helpfully reduces the variance of the
+estimator. First, rather than averaging samples $\tilde{y}_n^{(m)}$,
+the same result is obtained by averaging linear predictions,
+\begin{eqnarray*}
+\hat{\tilde{y}}_n
+& = & \mathbb{E}\left[
+          \alpha + \beta \cdot \tilde{x}_n
+          \mid \tilde{x}_n, x, y
+       \right]
+\\[4pt]
+& \approx &
+\frac{1}{M} \sum_{m = 1}^M
+  \alpha^{(m)} + \beta^{(m)} \cdot \tilde{x}_n
+\end{eqnarray*}
+This is possible because
+$$
+\tilde{y}_n^{(m)} \sim \textrm{normal}(\tilde{y}_n \mid \alpha^{(m)} +
+\beta^{(m)} \cdot \tilde{x}_n, \sigma^{(m)}),
+$$
+and the normal distribution has symmetric error so that the expectation of
+$\tilde{y}_n^{(m)}$ is the same as $\alpha^{(m)} + \beta^{(m)} \cdot
+\tilde{x}_n$.  Replacing the sampled quantity $\tilde{y}_n^{(m)}$ with
+its expectation is a general variance reduction technique for Monte
+Carlo estimates known as *Rao-Blackwellization* [@Rao:1945; @Blackwell:1947].
+
+In the linear case, because the predictor is linear in the
+coefficients, the estimate can be further simplified to use the
+estimated coefficients,
+\begin{eqnarray*}
+\tilde{y}_n^{(m)}
+& \approx &
+\frac{1}{M} \sum_{m = 1}^M
+  \left( \alpha^{(m)} + \beta^{(m)} \cdot \tilde{x}_n \right)
+\\[4pt]  
+& = & \frac{1}{M} \sum_{m = 1}^M \alpha^{(m)}
+      + \frac{1}{M} \sum_{m = 1}^M (\beta^{(m)} \cdot \tilde{x}_n)
+\\[4pt]  
+& = & \frac{1}{M} \sum_{m = 1}^M \alpha^{(m)}
+      + \left( \frac{1}{M} \sum_{m = 1}^M \beta^{(m)}\right) \cdot \tilde{x}_n
+\\[4pt]
+& = & \hat{\alpha} + \hat{\beta} \cdot \tilde{x}_n.
+\end{eqnarray*}
+
+In Stan, only the first of the two steps (the important variance
+reduction step) can be coded in the object model.  The linear
+predictor is defined in the generated quantities block.
+```
+data {
+  int<lower = 0> N_tilde;
+  vector[N_tilde] x_tilde;
+  ...
+generated quantities {
+  vector[N_tilde] tilde_y = alpha + beta * x_tilde;
+```
+The posterior mean of `tilde_y` calculated by Stan is the Bayesian
+estimate $\hat{\tilde{y}}.$  The posterior median may also be
+calculated and used as an estimate, though square error and the
+posterior mean are more commonly reported.
+
+## Cross-validation
+
+Cross-validation involves choosing multiple subsets of a data set as
+the test set and using the other data as training.  This can be done
+by partitioning the data and using each subset in turn as the test set
+with the remaining subsets as training data.  A partition into ten
+subsets is common to reduce computational overhead.  In the limit,
+when the test set is just a single item, the result is known as
+leave-one-out (LOO) cross-validation.
+
+Partitioning the data and reusing the partitions is very fiddly in the
+indexes and may not lead to even divisions of the data. It's far
+easier to use random partitions, which support arbitrarily sized
+test/training splits and can be easily implemented in Stan.
```

Comment:
Maybe add also an example code for cross-validation with permutation given as data?

Reply:
Thanks for letting me know.  I added discussion.  

I'll just mention you can take the permutation in as data as easily as generating it by a function.  I'm not sure why you'd want to do that rather than just doing the test/training split on the outside if you want to control it.  

I was going to write the K-fold version, but it's hard to write the code in general when the data might not be divisible by K.  I could add it.

Ack---I also realize I completely failed to mention the loo package and the approach it takes to doing approximate LOO x-validation.  I'll do that, too, but with just a citation, not a lot of detail.  It'd probably be good to expand later to talk more about approximate LOO.